jscomp/ml/typecore.ml

```diff
@@ -798,10 +801,36 @@ end) = struct
     | Tconstr(p, _, _) -> p
     | _ -> assert false
 
-  let lookup_from_type env tpath lid =
+  let lookup_from_type env tpath (lid : Longident.t loc) : Name.t =
     let descrs = get_descrs (Env.find_type_descrs tpath env) in
     Env.mark_type_used env (Path.last tpath) (Env.find_type tpath env);
-    match lid.txt with
+    let is_dict = Path.same tpath Predef.path_dict in
+    if is_dict then (
+      (* [dict] Dicts are implemented as a record with a single "magic" field. This magic field is 
+         used to track the dict value type, and any label lookup on the dict record type 
+         will give that single value type back. This is how we can piggy back on the record 
+         pattern matching mechanism.
+        
+        The code below handles directing any label lookup to the magic field. *)
+      match lid.txt with
+        Longident.Lident s_ -> begin
+          let s = 
+            if List.exists (fun nd -> get_name nd = s_) descrs
+            || not (List.exists (fun nd -> get_name nd = Dict_type_helpers.dict_magic_field_name) descrs)
+            then s_
+            else Dict_type_helpers.dict_magic_field_name in 
+          try
+            let x = List.find (fun nd -> get_name nd = s) descrs in
```

Comment:
find_opt cleaner perhaps?

Reply:
I think this whole section can be rewritten/simplified, because we no longer have the constraints we had initially.